### ISSUE TITLE
Multinomial (GPU ONLY) without replacement generates repeated items

### ISSUE THREAD

**Issue by xforceco** [pytorch/pytorch#16709]
## 🐛 Bug

Use multinomial on cuda without replacement, it generates repeated items occasionally.

## To Reproduce

```python
for i in range(10000):
    for j in torch.ones(600,30000,device="cuda").multinomial(1000):
        if j.size()!=j.unique().size():
            print('error')
```

## Expected behavior

That error shall never be printed out, but it did.

## Environment

python 3.6, pytorch 1.0.0, anaconda3, CUDA 9.1
Let me know if you cannot reproduce it

## Additional context

Spending a whole day to debug my own project code, thinking of underflow or overflow or any of my fault. It turns out the GPU version of multinomial is just bs, feels bad man.

**Comment by soumith** [comment]
I'm sorry you spent a lot of time hunting down to a multinomial bug, that's nasty.

I can reproduce it both on 1.0.0 and on master, we'll look into it with high priority.

**Comment by soumith** [comment]
this has come up in the past in https://github.com/pytorch/pytorch/issues/4858 and the comments there detail the issue, which is related to precision of some of the operations in there, such as cumsum.
PR https://github.com/pytorch/pytorch/pull/5774 was sent by @t-vi with investigation by @rfejgin to make the probability of such events happening really low (it was tested at 1 in 1 million), but maybe the underlying precision issue still exists. This PR is already in v1.0.0, so it didn't resolve your issue.

**Comment by t-vi** [comment]
Does #16075 not fix it either?

Am 4. Februar 2019 05:12:23 MEZ schrieb Soumith Chintala <notifications@github.com>:

**Comment by soumith** [comment]
@t-vi from what I can see, it doesnt fix it, because the problem exists on master.

**Comment by t-vi** [comment]
I would have hoped that after that you're only seeing 0 being returned too often.

**Comment by t-vi** [comment]
Running without restarting had me go past 400 without hitting the bug.
I would expect that we see excessive 0s - which is the same returned when you request more samples than you have categories with nonzero probabilites.

I was going to rewrite multinomial for speed improvements (see #13018) in ATen after @syed-ahmed 's big probability rewrite. We could have a global unique to see if anything is duplicate. Or you we could make the erroneous values -1 and re-sample them if they occur.

In the end this is one of the things where I really wished there was some error signalling softer than device assert (by "time"stamping tensors and errors and invalidating those past the error or so or even just throwing an exception on cuda sync but keeping the tensors around).

**Comment by t-vi** [comment]
So I ran Soumith's script on my branch of #14957 (which has a ~1 week old version of master) and it didn't find anything in 10000 seeds (took ~8 hours on my GTX1080Ti).
@ptrblck also ran it for a while (thanks!).

**Comment by soumith** [comment]
So, I reran it in a freshly compiled maste (and tried a freshly compiled v1.0.1), and cannot repro it anymore.
I suspect that my environment was hoarding a previous pytorch install without @t-vi 's fix.

@xforceco closing this as fixed via  #16075 which will be part of the v1.0.1 release, and is also fixed on master.
